src/dynd/string.cpp

```diff
@@ -5,14 +5,358 @@
 
 #include <dynd/string.hpp>
 #include <dynd/kernels/string_concat_kernel.hpp>
+#include <dynd/kernels/string_count_kernel.hpp>
+#include <dynd/kernels/string_find_kernel.hpp>
+#include <dynd/kernels/string_replace_kernel.hpp>
 #include <dynd/func/elwise.hpp>
 
 using namespace std;
 using namespace dynd;
 
+
+////////////////////////////////////////////////////////////
+// String algorithms
+
+
+void dynd::string_concat(size_t nop, dynd::string *d, const dynd::string *const *s)
+{
+  // Get the size of the concatenated string
+  size_t size = 0;
+  for (size_t i = 0; i != nop; ++i) {
+    size += (s[i]->end() - s[i]->begin());
+  }
+
+  // Allocate the output
+  d->resize(size);
+  // Copy the string data
+  char *dst = d->begin();
+  for (size_t i = 0; i != nop; ++i) {
+    size_t op_size = (s[i]->end() - s[i]->begin());
+    DYND_MEMCPY(dst, s[i]->begin(), op_size);
+    dst += op_size;
+  }
+}
+
+
+template<class SelfType>
+struct string_search_base {
+  inline void bloom_add(uint64_t &mask, const char ch)
+  {
+    mask |= 1UL << ((ch) & (64 - 1));
+  }
+
+  inline uint64_t bloom(const uint64_t &mask, const char ch)
+  {
+    return mask & (1UL << ((ch) & (64 - 1)));
+  }
+
+  /* Special case path where the needle is only a single character.
+     According to effbot's numbers, for haystacks < 10 characters, a
+     naive loop is fastest.  Larger than that, using POSIX `memchr` is
+     faster. */
+  inline void find_1char(const char *haystack, size_t n, char needle)
+  {
+    SelfType *self = (SelfType *)this;
+
+    if (n <= 10) {
+      for (size_t i = 0; i < n; ++i) {
+        if (haystack[i] == needle) {
+          if (self->handle_match(i)) {
+            return;
+          }
+        }
+      }
+    } else {
+      const char *s = haystack;
+      while (s < haystack + n) {
+        void *candidate = memchr((void *)s, needle, n);
+        if (candidate == NULL) {
+          return;
+        }
+        s = (const char *)candidate;
+        if (self->handle_match(s - haystack)) {
+          return;
+        }
+        s++;
+      }
+    }
+  }
+
```

Comment:
`inline` is actually implicit for any methods in a class or struct, so you can drop it.